tests/xharness/TestReporter.cs

```diff
@@ -0,0 +1,490 @@
+				if (idx > 0) {
+					idx += str.Length;
+					var next_idx = line.IndexOf ('\'', idx);
+					if (next_idx > idx)
+						int.TryParse (line.Substring (idx, next_idx - idx), out pid);
+				}
+				if (pid != -1)
+					return pid;
+			}
+			return pid;
+		}
+
+		// return the reason for a crash found in a log
+		void GetCrashReason (int pid, ILog crashLog, out string crashReason)
+		{
+			crashReason = null;
+			using var crashReader = crashLog.GetReader (); // dispose when we leave the method
+			var text = crashReader.ReadToEnd ();
+
+			var reader = JsonReaderWriterFactory.CreateJsonReader (Encoding.UTF8.GetBytes (text), new XmlDictionaryReaderQuotas ());
+			var doc = new XmlDocument ();
+			doc.Load (reader);
+			foreach (XmlNode node in doc.SelectNodes ($"/root/processes/item[pid = '" + pid + "']")) {
+				Console.WriteLine (node?.InnerXml);
+				Console.WriteLine (node?.SelectSingleNode ("reason")?.InnerText);
+				crashReason = node?.SelectSingleNode ("reason")?.InnerText;
+			}
+		}
+
+		// return if the tcp connection with the device failed
+		async Task<bool> TcpConnectionFailed ()
+		{
+			using var reader = new StreamReader (runner.MainLog.FullPath);
+			string line;
+			while ((line = await reader.ReadLineAsync ()) != null) {
+				if (line.Contains ("Couldn't establish a TCP connection with any of the hostnames")) {
+					return true;
+				}
+			}
+			return false;
+		}
+
+		// kill any process 
+		Task KillAppProcess (int pid, CancellationTokenSource cancellationSource) { 
+				var launchTimedout = cancellationSource.IsCancellationRequested;
+				var timeoutType = launchTimedout ? "Launch" : "Completion";
+				var timeoutValue = launchTimedout ? runner.LaunchTimeout : Timeout.TotalSeconds;
+
+				runner.MainLog.WriteLine ($"{timeoutType} timed out after {timeoutValue} seconds");
+				return ProcessManager.KillTreeAsync (pid, runner.MainLog, true);
+		}
+
+		async Task CollectResult (Task<ProcessExecutionResult> processExecution)
+		{ 
+			// wait for the execution of the process, once that is done, perform all the parsing operations and
+			// leave a clean API to be used by AppRunner, hidding all the diff details
+			var result = await processExecution;
+			if (!waitedForExit && !result.TimedOut) {
+				// mlaunch couldn't wait for exit for some reason. Let's assume the app exits when the test listener completes.
+				runner.MainLog.WriteLine ("Waiting for listener to complete, since mlaunch won't tell.");
+				if (!await listener.CompletionTask.TimeoutAfter (Timeout - TimeoutWatch.Elapsed)) {
+					result.TimedOut = true;
+				}
+			}
+
+			if (result.TimedOut) {
+				timedout = true;
+				Success = false;
+				runner.MainLog.WriteLine (timeoutMessage, Timeout.TotalMinutes);
+			} else if (result.Succeeded) {
+				runner.MainLog.WriteLine (completionMessage);
+				Success = true;
+			} else {
+				runner.MainLog.WriteLine (failureMessage);
+				Success = false;
+			}
+		}
+
+		public void LaunchCallback (Task<bool> launchResult)
```

Comment:
Moved the logic from AppRunner here, since it is more important for the TestReporter.